docs/configuration.md
# Configuration Guide

## Summary

This SDK uses a structure called "Config" to store and manage configuration, read ["config/config.go"](https://github.com/peta-express/cloud-sdk-go/blob/master/config/config.go) file's comments of public functions for more information.

Except for AccessKeyID and SecretAccessKey, you can also configure the API servers for private cloud usage scenario. All available configurable items are list in default configuration file.

___Default Configuration File:___

``` yaml
# PetaExpress services configuration

qy_access_key_id: 'ACCESS_KEY_ID'
qy_secret_access_key: 'SECRET_ACCESS_KEY'

host: 'api.petaexpress.com'
port: 443
protocol: 'https'
uri: '/iaas'
connection_retries: 3

# Valid log levels are "debug", "info", "warn", "error", and "fatal".
log_level: 'warn'

```

## Usage

1. Just create a config structure instance with your API AccessKey, and initialize services that you need using Init() function of target service.

2. Or if you do not want to expose your AccessKeyID and SecretAccessKey, you can also call our API without them in the following way:
- Go to our IAM service, create an instance role and attach it to your instance.
- Create the config structure instance without AccessKeyID and SecretAccessKey.

Note that you can also configure your own credential proxy server (where you retrieve tokens) in configuration file like this:

```yaml
credential_proxy_protocol: 'http'
credential_proxy_host: '169.254.169.254'
credential_proxy_port: 80
credential_proxy_uri: '/latest/meta-data/security-credentials'
```

### Code Snippet

Create default configuration

``` go
defaultConfig, _ := config.NewDefault()
```

Create configuration from AccessKey

``` go
configuration, _ := config.New("ACCESS_KEY_ID", "SECRET_ACCESS_KEY")

anotherConfiguration := config.NewDefault()
anotherConfiguration.AccessKeyID = "ACCESS_KEY_ID"
anotherConfiguration.SecretAccessKey = "SECRET_ACCESS_KEY"
```

Load user configuration

``` go
userConfig, _ := config.NewDefault().LoadUserConfig()
```

Load configuration from config file

``` go
configFromFile, _ := config.NewDefault().LoadConfigFromFilepath("PATH/TO/FILE")
```

Change API server

``` go
moreConfiguration, _ := config.NewDefault()

moreConfiguration.Protocol = "https",
moreConfiguration.Host = "api.private.com",
moreConfiguration.Port = 4433,
moreConfiguration.URI = "/iaas",
```


CHANGELOG.md
# Change Log
All notable changes to PetaExpress SDK for Go will be documented in this file.


config/config.go
// +-------------------------------------------------------------------------
// | Copyright (C) 2024 PEG Tech, Inc.
// +-------------------------------------------------------------------------
// | Licensed under the Apache License, Version 2.0 (the "License");
// | you may not use this work except in compliance with the License.
// | You may obtain a copy of the License in the LICENSE file, or at:
// |
// | http://www.apache.org/licenses/LICENSE-2.0
// |
// | Unless required by applicable law or agreed to in writing, software
// | distributed under the License is distributed on an "AS IS" BASIS,
// | WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// | See the License for the specific language governing permissions and
// | limitations under the License.
// +-------------------------------------------------------------------------

package config

import (
	"fmt"
	"io/ioutil"
	"net"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/peta-express/cloud-sdk-go/logger"
	"github.com/peta-express/cloud-sdk-go/utils"
)

// A Config stores a configuration of this sdk.
type Config struct {
	AccessKeyID     string `yaml:"qy_access_key_id"`
	SecretAccessKey string `yaml:"qy_secret_access_key"`

	Host              string `yaml:"host"`
	Port              int    `yaml:"port"`
	Protocol          string `yaml:"protocol"`
	URI               string `yaml:"uri"`
	ConnectionRetries int    `yaml:"connection_retries"`
	ConnectionTimeout int    `yaml:"connection_timeout"`

	LogLevel string `yaml:"log_level"`

	Zone string `yaml:"zone"`

	CredentialProxyProtocol string `yaml:"credential_proxy_protocol"`
	CredentialProxyHost     string `yaml:"credential_proxy_host"`
	CredentialProxyPort     int    `yaml:"credential_proxy_port"`
	CredentialProxyURI      string `yaml:"credential_proxy_uri"`

	Token      string
	Expiration int64

	Connection *http.Client
}

// New create a Config with given AccessKeyID and SecretAccessKey.
func New(accessKeyID, secretAccessKey string) (*Config, error) {
	config, err := NewDefault()
	if err != nil {
		return nil, err
	}

	config.AccessKeyID = accessKeyID
	config.SecretAccessKey = secretAccessKey

	config.Connection = &http.Client{}

	return config, nil
}

// NewWithEndpoint create a Config with given AccessKeyID, SecretAccessKey and endpoint
func NewWithEndpoint(accessKeyID, secretAccessKey, endpoint string) (*Config, error) {
	qcURL, err := url.Parse(endpoint)
	if err != nil {
		return nil, err
	}
	if qcURL.Opaque != "" {
		return nil, fmt.Errorf("wrong URL format")
	}

	if !strings.Contains(qcURL.Host, ":") {
		if qcURL.Scheme == "https" {
			qcURL.Host += ":443"
		} else if qcURL.Scheme == "http" {
			qcURL.Host += ":80"
		} else {
			return nil, fmt.Errorf("can not find default port ")
		}
	}

	config, err := NewDefault()
	if err != nil {
		return nil, err
	}

	host, port, err := net.SplitHostPort(qcURL.Host)
	if err != nil {
		return nil, err
	}

	config.Port, _ = strconv.Atoi(port)
	config.Host = host
	config.AccessKeyID = accessKeyID
	config.SecretAccessKey = secretAccessKey
	config.Protocol = qcURL.Scheme
	config.URI = qcURL.Path
	config.Connection = &http.Client{}
	return config, nil
}

// NewDefault create a Config with default configuration.
func NewDefault() (*Config, error) {
	config := &Config{}
	err := config.LoadDefaultConfig()
	if err != nil {
		return nil, err
	}

	timeout := time.Duration(config.ConnectionTimeout) * time.Second
	transport := &http.Transport{
		Dial: func(network, addr string) (net.Conn, error) {
			return net.DialTimeout(network, addr, timeout)
		},
	}
	config.Connection = &http.Client{
		Transport: transport,
	}

	return config, nil
}

// LoadDefaultConfig loads the default configuration for Config.
// It returns error if yaml decode failed.
func (c *Config) LoadDefaultConfig() error {
	_, err := utils.YAMLDecode([]byte(DefaultConfigFileContent), c)
	if err != nil {
		logger.Error("Config parse error: " + err.Error())
		return err
	}

	logger.SetLevel(c.LogLevel)

	return nil
}

// LoadUserConfig loads user configuration in ~/.petaexpress/config.yaml for Config.
// It returns error if file not found.
func (c *Config) LoadUserConfig() error {
	_, err := os.Stat(GetUserConfigFilePath())
	if err != nil {
		logger.Warn("Installing default config file to \"" + GetUserConfigFilePath() + "\"")
		InstallDefaultUserConfig()
	}

	return c.LoadConfigFromFilepath(GetUserConfigFilePath())
}

// LoadConfigFromFilepath loads configuration from a specified local path.
// It returns error if file not found or yaml decode failed.
func (c *Config) LoadConfigFromFilepath(filepath string) error {
	if strings.Index(filepath, "~/") == 0 {
		filepath = strings.Replace(filepath, "~/", getHome()+"/", 1)
	}

	configYAML, err := ioutil.ReadFile(filepath)
	if err != nil {
		logger.Error("File not found: " + filepath)
		return err
	}

	return c.LoadConfigFromContent(configYAML)
}

// LoadConfigFromContent loads configuration from a given byte slice.
// It returns error if yaml decode failed.
func (c *Config) LoadConfigFromContent(content []byte) error {
	c.LoadDefaultConfig()

	_, err := utils.YAMLDecode(content, c)
	if err != nil {
		logger.Error("Config parse error: " + err.Error())
		return err
	}

	logger.SetLevel(c.LogLevel)

	timeout := time.Duration(c.ConnectionTimeout) * time.Second
	transport := &http.Transport{
		Dial: func(network, addr string) (net.Conn, error) {
			return net.DialTimeout(network, addr, timeout)
		},
	}
	c.Connection = &http.Client{
		Transport: transport,
	}

	return nil
}
